js/ai/src/metrics.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Histogram, Meter, metrics, ValueType } from '@opentelemetry/api';
+import { GenerateResponse } from './generate/response.js';
+
+type MetricCreateFn<T> = (meter: Meter) => T;
+export const METER_NAME = 'genkit';
+
+/**
+ * Wrapper for OpenTelemetry metrics.
+ *
+ * The OpenTelemetry {MeterProvider} can only be accessed through the metrics
+ * API after the NodeSDK library has been initialized. To prevent race
+ * conditions we defer the instantiation of the metric to when it is first
+ * ticked.
+ */
+class Metric<T> {
```

Comment:
Not a huge fan of having this be in core - are we positive that this can't live in the plugin?

Reply:
There's no reason it couldn't be done in the plugin, but these semantic conventions are intended so that different tools can understand the telemetry data, so only exporting them when you're using GCP somewhat defeats the purpose.